doc/bgp_error_handling/BGP_Route_Error_Handling_Arlo.md

```diff
@@ -0,0 +1,241 @@
+
+# Definition/Abbreviation
+### Table 1: Abbreviations
+
+| **Term**          | ***Meaning***           |
+|-------------------|-------------------------|
+| BGP               | Border Gateway Protocol |
+| GR                | Graceful Restart        |
+| SONIC             | Software for Open Networking in the Cloud                        |
+| FRR    	        |  FRRouting                       |
+| FPM               |  Forwarding Plane Manager        |
+| SwSS              |  SONiC Switch State Service      |
+# 1 Requirement Overview
+ When BGP learns a prefix, it sends it to route table manager(Zebra). The routes are installed in kernel and sent to APP_DB via fpmsyncd. 
+The Orchagent reads the route from APP_DB, creates new resources like nexthop or nexthop group Id and installs the route in ASIC_DB. The syncd triggers the appropriate SAI API and route is installed in hardware. The CRM manages the count of critical resources allocated by orchagent through SAI API.
+Due to resource allocations failures in hardware, SAI API calls can fail and these failures should be notified to Zebra and BGP.
+On learning the prefix, BGP can immediately advertise the prefix to its neighbors. However, if the error-handling feature is enabled, BGP waits for success notification from hardware before advertising the same to its peers. If the hardware returns error, the routes are not advertised to the peers. 
+
+## 1.1 Functional Requirements
+
+ 
+
+ 1. BGP should not advertise the routes which have failed to be installed in hardware.
+ 1. BGP should mark the routes which are not installed in hardware as  "FIB-install pending" routes in its RIB-IN table.
+ 1. Zebra should mark the routes which are not successfully installed in hardware as failed routes.
+
+## 1.2 Configuration and Management Requirements
+## 1.3 Scalability Requirements
+## 1.4 Warm Boot Requirements
+   There is no change needed in BGP warm reboot for supporting this feature.
+# 2 Functionality
+Refer to section 1
+
+## 2.1 Target Deployment Use Cases
+
+## 2.2 Functional Description
+Refer to section 1.1
+
+# 3 Design
+## 3.1 Overview
+On enabling the error-handling feature, fpmsyncd subscribes to the changes in the ERROR_ROUTE_TABLE entries. Whenever the error status in ERROR_ROUTE_TABLE is updated, fpmsyncd is notified. It then sends a message to Zebra's routing table to take appropriate action.
+Zebra should lookup the route and mark it as not installed in hardware. It should create a route-netlink message to withdraw this state in kernel. Also, it sends message to the source protocol for the route (BGP). 
+BGP marks the route as not installed in hardware and does not advertise the route to its peers. It should mark the route in RIB-IN as not installed in hardware and remove it from RIB-OUT list, if any. 

```

Comment:
Can this feature turn-off on demand ? is yes, can this affect the system stability?

Reply:
This feature can be turned off on demand. But it can affect the system stability. When the config was turned on, there may be some routes in BGP, for which, it is waiting for update from hardware. Suddenly, the feature is turned off. In this case, we will unsubscribe from ERROR_DB and will no longer receive any notifications from hardware. Hence, some of the routes may not receive any notification from hardware. 
It is recommended to restart the BGP docker when the config state is changed to disable from enable. By default, this config is disabled. If the config is changed from disable to enable, we do not need to restart the docker. But the feature will be affecting only those routes which will be learnt after enabling the feature.